File: scripts/build-book-data.mjs
#!/usr/bin/env node
import fs from 'node:fs';import path from 'node:path';
const root=process.cwd();const bookId=process.argv[2]||'shanghai-g5a-2026';
const dir=path.join(root,'books',bookId);const book=JSON.parse(fs.readFileSync(path.join(dir,'book.json'),'utf8'));const lessons=JSON.parse(fs.readFileSync(path.resolve(root,book.lessons),'utf8'));
const stamp=(s)=>{const m=Math.floor(s/60),sec=s-m*60;return `[${String(m).padStart(2,'0')}:${sec.toFixed(3).padStart(6,'0')}]`;};
let built=0;
for(const lesson of lessons){if(!lesson.timing||!lesson.transcript)continue;const timing=JSON.parse(fs.readFileSync(path.resolve(root,lesson.timing),'utf8'));const lines=[`[al:${book.title}]`,`[ar:Original Publisher Audio]`,`[ti:${lesson.title}]`,`[by:alignment=${timing.alignmentSource||lesson.alignmentSource||'unknown'}]`,''];for(const s of timing.segments||[])lines.push(`${stamp(Number(s.start))}${s.en}${s.cn?'|'+s.cn:''}`);const out=path.resolve(root,lesson.transcript);fs.mkdirSync(path.dirname(out),{recursive:true});fs.writeFileSync(out,lines.join('\n')+'\n');console.log('BUILD',lesson.id,'->',path.relative(root,out));built++;}
console.log(`Built ${built} transcript(s).`);

File: scripts/build-book-data.py
#!/usr/bin/env python3
"""Generate/synchronize LRC files from explicit segment timing JSON.

This keeps the data pipeline reproducible: edit the structured segment JSON,
then regenerate compatible LRC rather than hand-copying timestamps.
"""
from __future__ import annotations
import argparse, json
from pathlib import Path

def fmt(t):
    t=float(t); m=int(t//60); s=t-m*60
    return f'[{m:02d}:{s:06.3f}]'

def build(timing_path:Path, out:Path, album:str, artist:str, title:str, by:str='book-data-builder'):
    data=json.loads(timing_path.read_text('utf-8'))
    segs=data.get('segments',[])
    if not segs: raise SystemExit(f'No segments in {timing_path}')
    lines=[f'[al:{album}]',f'[ar:{artist}]',f'[ti:{title}]',f'[by:{by}]']
    last=-1
    for i,s in enumerate(segs,1):
        start=float(s['start']); end=float(s['end']); text=str(s.get('text','')).strip()
        if not text: raise SystemExit(f'Empty text segment {i}')
        if start <= last: raise SystemExit(f'Non-increasing start at segment {i}')
        if end <= start: raise SystemExit(f'Invalid end at segment {i}')
        last=start
        cn=str(s.get('cn','')).strip()
        lines.append(f'{fmt(start)}{text}|{cn}')
    out.parent.mkdir(parents=True,exist_ok=True)
    out.write_text('\n'.join(lines)+'\n',encoding='utf-8')
    print(f'WROTE {out} ({len(segs)} lines)')

def main():
    ap=argparse.ArgumentParser()
    ap.add_argument('--timings',required=True)
    ap.add_argument('--out',required=True)
    ap.add_argument('--album',required=True)
    ap.add_argument('--artist',default='Publisher Original Audio')
    ap.add_argument('--title',required=True)
    ap.add_argument('--by',default='book-data-builder')
    args=ap.parse_args()
    build(Path(args.timings),Path(args.out),args.album,args.artist,args.title,args.by)

if __name__=='__main__': main()
